### PR TITLE
honeybadger processing errors

### PR DESCRIPTION
**Pull request by ncliffor**
https://www.pivotaltracker.com/n/projects/319615/stories/96551200

### CODE REVIEW

**Comment by ncliffor** [comment]
@ericsullivan quick attempt at gathering processing errors with honeybadger. Does something like this make sense? Or should it be kept in billing?

File: lib/dalliance.rb [hunk 1]
```diff
@@ -21,7 +21,8 @@ def options
         :worker_class => detect_worker_class,
         :queue => 'dalliance',
         :logger => detect_logger,
-        :duration_column => 'dalliance_duration'
+        :duration_column => 'dalliance_duration',
+        :honeybadger_api_key => (defined?(Rails) ? honeybadger_api_key : nil)
```
Comment (ncliffor) [inline lib/dalliance.rb]: set in `initializers/dalliance.rb` in fms_billing_client

**Comment by ncliffor** [comment]
Also not sure if this will work the way I'm thinking, just wanted to run the thought by you

**Comment by ericsullivan** [comment]
it's an ok idea, but this is a specific implementation. Do an adapter pattern, like, call error_notification(error) and let the client do something with that. Then, in the client, you can send to honeybadger or whatever...

**Comment by ncliffor** [comment]
@ericsullivan, does something like this look better with `report_error = HoneyBadger.notify` in billing initializer? Not sure if its exactly right or not...

File: lib/dalliance.rb [hunk 2]
```diff
@@ -245,6 +246,10 @@ def dalliance_process(backgound_processing = false)
         self.class.where(id: self.id).update_all(self.class.dalliance_options[:duration_column] => duration.to_i)
       end
     end
+
+    if errors.present?
+      Dalliance::ErrorNotification.new.notify(errors)
```
Comment (ericsullivan) [inline lib/dalliance.rb]: if dalliance.config.error_notifier
  error_notifier.notify(errors)
end

leave the implementation up to the user

Comment (ncliffor) [inline lib/dalliance.rb]: ahh, ok, I think I got it now. I'm assuming the ErrorNotification stuff is unnecessary then?

Comment (ericsullivan) [inline lib/dalliance.rb]: Correct

**Comment by ncliffor** [comment]
@ericsullivan @GregoryDuguies @seanbjornsson

File: lib/dalliance.rb [hunk 3]
```diff
@@ -245,6 +254,10 @@ def dalliance_process(backgound_processing = false)
         self.class.where(id: self.id).update_all(self.class.dalliance_options[:duration_column] => duration.to_i)
       end
     end
+
+    if errors.present? && error_notifier.present?
```
Comment (ericsullivan) [inline lib/dalliance.rb]: move this up to the actual error code (line 241)
error_notifier.call(e) if error_notifier.present?

**Comment by seanbjornsson** [comment]
a passing thought: is there any reason we would want to support an error handler/error response that could do something like this?
```ruby
->(error_hash){
  Honeybadger.notify(error_hash[:error])
  case error_hash[:calling_klass]
  when /Report/, /Export/
    # do something that you might want to do when reports have a processing error
  when /SynchronousEligibility/
    # do something that you want to do when HTS has a processing error
  end
}
```

**Comment by ericsullivan** [comment]
I'll leave that up to you. Passing e is simpler. passing e and a class?

**Comment by seanbjornsson** [comment]
i agree. i couldnt think of how to pass the class off the top of my head. and if we dont actually need it, then its easier to keep it simple

**Comment by ericsullivan** [comment]
then keep it simple :)

On Wed, Jul 22, 2015 at 10:06 AM, seanbjornsson <notifications@github.com>
wrote:

> i agree. i couldnt think of how to pass the class off the top of my head.
> and if we dont actually need it, then its easier to keep it simple
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/annkissam/dalliance/pull/3#issuecomment-123734068>.
>



-- 
Eric Sullivan

Annkissam: Mission Driven Systems
www.annkissam.com

One Broadway, 14th Floor
Cambridge, MA 02142


C. 978-376-2543 *preferred
B. 617-401-2480
F. 617-507-5922

**Comment by ncliffor** [comment]
@ericsullivan, should we update the version.rb for this?

**Comment by ericsullivan** [comment]
I can bump the version, but I'd like a spec to confirm it's working

On Mon, Jul 27, 2015 at 12:09 PM, Nick Clifford <notifications@github.com>
wrote:

> @ericsullivan <https://github.com/ericsullivan>, should we update the
> version.rb for this?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/annkissam/dalliance/pull/3#issuecomment-125258004>.
>



-- 
Eric Sullivan

Annkissam: Mission Driven Systems
www.annkissam.com

One Broadway, 14th Floor
Cambridge, MA 02142


C. 978-376-2543 *preferred
B. 617-401-2480
F. 617-507-5922

**Comment by ncliffor** [comment]
@ericsullivan sorry for the delay. Let me know if this looks good or if you had something else in mind :)

File: lib/dalliance.rb [hunk 4]
```diff
@@ -213,6 +222,8 @@ def dalliance_process(backgound_processing = false)
       #Save the error for future analysis...
       self.dalliance_error_hash = {:error => e.class.name, :message => e.message, :backtrace => e.backtrace}
 
+      error_notifier.call(e) if error_notifier.present?
+
```
Comment (ericsullivan) [inline lib/dalliance.rb]: why is this here twice?

File: spec/dalliance/synchronous_spec.rb [hunk 5]
```diff
@@ -75,6 +75,17 @@
       expect(subject.dalliance_error_hash).not_to be_empty
       expect(subject.dalliance_error_hash[:error]).to eq('Persistance Failure: See Logs')
     end
+
+    context "error_notifier" do
+      it "should pass the errors" do
+        DallianceModel.dalliance_options[:error_notifier] = ->(error){ @error_report = "#{error}" }
+        allow_any_instance_of(DallianceModel).to receive(:error_dalliance!).and_raise(RuntimeError.new)
```
Comment (ericsullivan) [inline spec/dalliance/synchronous_spec.rb]: remove this line and treat it like a standard error

**Comment by ncliffor** [comment]
@ericsullivan for the synchronous and asynchronous specs the error_notifier is reporting nil from dalliance.rb line 244. Is this because dalliance is still in the background process, or am I reading the code incorrectly?

**Comment by ncliffor** [comment]
just a heads up thats after removing the redundant error_notifier on line 225

File: spec/dalliance/asynchronous_delayed_job_spec.rb [hunk 6]
```diff
@@ -155,6 +155,18 @@
...
+
+    context "error_notifier" do
+      it "should pass the errors" do
+        DallianceModel.dalliance_options[:error_notifier] = ->(error){ @error_report = "#{error}" }
+
+        subject.dalliance_background_process
+        Delayed::Worker.new(:queues => [:dalliance]).work_off
+        subject.reload
+
+        expect(@error_report).to eq(nil) # Still processing, error not thrown yet...
```
Comment (ncliffor) [inline spec/dalliance/asynchronous_delayed_job_spec.rb]: @ericsullivan the error should be stored but not yet raised here, correct?

**Comment by ncliffor** [comment]
@ericsullivan bump

File: lib/dalliance.rb [hunk 7]
```diff
@@ -229,7 +246,7 @@ def dalliance_process(backgound_processing = false)
       end
 
       #Don't raise the error if we're backgound_processing...
-      raise e unless backgound_processing && self.class.dalliance_options[:worker_class].rescue_error?
```
Comment (ericsullivan) [inline lib/dalliance.rb]: don't change this, it's related to how background job / resque work. Just add another line (before it) where you call the error_notifiier. Also, the null object pattern could work nicely. Have the default error notifier be ->(e) {}, and then you can always call it

File: spec/dalliance/asynchronous_delayed_job_spec.rb [hunk 8]
```diff
@@ -155,6 +155,18 @@
...
       expect(subject.dalliance_error_hash[:error]).to eq('Persistance Failure: See Logs')
     end
+
+    context "error_notifier" do
+      it "should pass the errors" do
+        DallianceModel.dalliance_options[:error_notifier] = ->(error){ @error_report = "#{error}" }
+
+        subject.dalliance_background_process
+        Delayed::Worker.new(:queues => [:dalliance]).work_off
+        subject.reload
```
Comment (ericsullivan) [inline spec/dalliance/asynchronous_delayed_job_spec.rb]: Why do you need a reload?

File: spec/dalliance/asynchronous_delayed_job_spec.rb (hunk 6 above)
Comment (ericsullivan) [inline spec/dalliance/asynchronous_delayed_job_spec.rb]: it should be set

File: spec/dalliance/asynchronous_resque_spec.rb [hunk 9]
```diff
@@ -171,6 +171,18 @@
...
       expect(subject.dalliance_error_hash[:error]).to eq('Persistance Failure: See Logs')
     end
+
+    context "error_notifier" do
+      it "should pass the errors" do
+        DallianceModel.dalliance_options[:error_notifier] = ->(error){ @error_report = "#{error}" }
+
+        subject.dalliance_background_process
+        Resque::Worker.new(:dalliance).process
+        subject.reload
```
Comment (ericsullivan) [inline spec/dalliance/asynchronous_resque_spec.rb]: no reload

**Comment by ncliffor** [comment]
@ericsullivan, updates as per your comments. Thanks for the help!

**Comment by ericsullivan** [comment]
Nice, I'll push a new version

File: lib/dalliance.rb [hunk 10]
```diff
@@ -228,6 +237,8 @@ def dalliance_process(backgound_processing = false)
         end
       end
 
+      error_notifier.call(e)
```
Comment (GregoryDuguies) [inline lib/dalliance.rb]: this needs to be in background processing yes? or did we just choose dalliance_process

Comment (ericsullivan) [inline lib/dalliance.rb]: not following your question

Comment (GregoryDuguies) [inline lib/dalliance.rb]: @ericsullivan I thought this was an issue with not properly calling the error_notifier in a a dalliance_background_process, but Nick and I have actually realized that if a job is processing in the background with resque, the honeybadger notifier doesn't seem to be called or working properly. I'm not sure why the tests pass on dalliance but it doesn't work on the server. 

If you call the notifier directly in dalliance in the background processing method, the notification is sent to honeybadger, but if the object raises an error in the process part, it doesn't send the notification to honeybadger. Do you have any thoughts as to why?

Comment (ericsullivan) [inline lib/dalliance.rb]: Why don't we upgrade honeybadger and see if that fixes it?

https://www.honeybadger.io/documentation/queue_integration

It also means we remove the resque-honeybadger gems

On Mon, Sep 14, 2015 at 5:32 PM, Gregory Duguies <notifications@github.com>
wrote:

> In lib/dalliance.rb
> <https://github.com/annkissam/dalliance/pull/3#discussion_r39450111>:
>
> > @@ -228,6 +237,8 @@ def dalliance_process(backgound_processing = false)
> >          end
> >        end
> >
> > +      error_notifier.call(e)
>
> @ericsullivan <https://github.com/ericsullivan> I thought this was an
> issue with not properly calling the error_notifier in a a
> dalliance_background_process, but Nick and I have actually realized that if
> a job is processing in the background with resque, the honeybadger notifier
> doesn't seem to be called or working properly. I'm not sure why the tests
> pass on dalliance but it doesn't work on the server.
>
> If you call the notifier directly in dalliance in the background
> processing method, the notification is sent to honeybadger, but if the
> object raises an error in the process part, it doesn't send the
> notification to honeybadger. Do you have any thoughts as to why?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/annkissam/dalliance/pull/3/files#r39450111>.
>



-- 
Eric Sullivan

Annkissam: Mission Driven Systems
www.annkissam.com

One Broadway, 14th Floor
Cambridge, MA 02142


C. 978-376-2543 *preferred
B. 617-401-2480
F. 617-507-5922
